library/std/src/ffi/mod.rs

```diff
@@ -127,6 +127,42 @@
 //! trait, which provides a [`from_wide`] method to convert a native Windows
 //! string (without the terminating nul character) to an [`OsString`].
 //!
+//! ## On all platforms
+//!
+//! On all platforms, [`OsStr`] consists of a sequence of bytes; see [`OsString`] for
+//! more details on its encoding on different platforms.
+//!
+//! While the encoding is not fully specified for all platforms, some rules for interacting with
+//! the encoding in `unsafe` blocks are:
+//! - When [transmuting] from `&OsStr` to `&[u8]`, any bytes that cannot be represented as `&str`
+//!   must be treated as opaque and only comparable within the same rust version built for the same
+//!   target platform.
+//! - When [transmuting] from `&[u8]` to `&OsStr`,
+//!   - the slice may only include content from comparable `&OsStr` (see above) or be valid UTF-8
+//!   - any splits of the `&OsStr` must be along char boundaries (the first byte of a UTF-8 code
+//!     point sequence)
```

Comment:
I’m not sure "char boundary" is well-defined for byte sequences that are not well-formed in UTF-8, especially on Unix with completely arbitrary byte sequences. Something sufficient for WTF-8 might be "any split must not be immediately before a byte in the `0x80..=0xBF` range (an UTF-8 continuation byte)"